### PR TITLE
adding output cache tutorial for .NET 8

### PR DESCRIPTION
**Pull request by CawaMS**
## Purpose

* ...

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
+﻿using StackExchange.Redis;
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs:1]: Is this file needed?

File: tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+namespace OutputCacheDallESample
+{
+    public static class GenerateImageSC
+    {
+        private static RedisConnectionProvider _provider;
+        public static async Task GenerateImageSCAsync(HttpContext context, string _prompt, IConfiguration _config)
+        {
+            string imageURL;
+
+            string endpoint = _config["AZURE_OPENAI_ENDPOINT"];
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs:16]: Please fix potential warnings here , also other places:
> Converting null literal or possible null value to non-nullable type.

File: tutorial/output-cache-open-ai/infra/abbreviations.json [hunk 3]
```diff
@@ -0,0 +1,135 @@
+{
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/infra/abbreviations.json:1]: what is this file used for?

File: tutorial/output-cache-open-ai/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# .NET 8 Web Application with Redis Output Caching and Azure Open AI
+
+## Overview
+This repository contains sample code for an AI image generation web application. It calls into Azure Open AI DallE to generate images based on user prompt. There are two features the sample highlights:
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/README.md:4]: Should be `this folder`?

File: tutorial/output-cache-open-ai/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+* It allows responses from the /cached/ endpoint to be saved in Azure Cache for Redis through the [IOutputCache() abstraction](https://learn.microsoft.com/aspnet/core/performance/caching/output?preserve-view=true&view=aspnetcore-8.0#cache-storage). 
+* It uses semantic caching to cache responses from similar prompts using the [Redis OM for .NET](https://github.com/redis/redis-om-dotnet)
+
+To experiment with the demo:
+1. call the /nocache/ endpoint with your command prompt. Refresh the browser. Observe that refresh takes a noticeable delay and a new picture will be generated.
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/README.md:9]: Could you provide an example command of how to `call the /nocache/` through command prompt? I would expect a url with `/noncache/` to load in the browser?

File: tutorial/output-cache-open-ai/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Prerequisites:
+- Azure subscription. [Start free](https://azure.microsoft.com/free)
+- .NET 8 or above. [Download](https://dotnet.microsoft.com/download/dotnet/8.0)
+- Docker. [Get docker](https://docs.docker.com/get-docker/)
+- Azure Developer CLI. [Install](https://learn.microsoft.com/azure/developer/azure-developer-cli/install-azd?tabs=winget-windows%2Cbrew-mac%2Cscript-linux&pivots=os-windows)
+
+1. Open a command prompt
+2. Change directory to the project folder where azure.yaml file is located
+4. Make sure docker is running
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/README.md:37]: Please provider command of how to cross check docker is running?

File: tutorial/output-cache-open-ai/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+5. Run:
+    ```
+    azd up
+    ```
+6. Follow command prompt to enter environment name and select subscription
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/README.md:42]: Does a user need some registration to OpenAI?

File: tutorial/output-cache-open-ai/infra/shared/apps-env.bicep [hunk 8]
```diff
@@ -0,0 +1,33 @@
+param name string
+param location string = resourceGroup().location
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/infra/shared/apps-env.bicep:2]: If we onboard new tutorials requiring the same infra setup, let us move this folder to a common place to avoid duplicate.

File: tutorial/output-cache-open-ai/OutputCacheDallESample/Program.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+builder.Services.AddStackExchangeRedisOutputCache(options => {
+    options.Configuration = builder.Configuration["RedisCacheConnection"];
+});
+builder.Services.AddOutputCache(options => {
+    // optional: named output-cache profiles
+});
+
+var app = builder.Build();
+
+app.MapGet("/", async (HttpContext context) => { await context.Response.WriteAsync("<h1>Welcome to OpenAI Art Gallery</h1>"); });
```
Comment (philon-msft) [inline tutorial/output-cache-open-ai/OutputCacheDallESample/Program.cs:18]: Semicolons and curly brackets aren't needed for single-line `Func`s. Also please format all `MapGet()` calls consistently

File: tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs (hunk 1 above)
Comment (philon-msft) [reply tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs:1]: I'd vote "no" - this ForceReconnect pattern is not needed with current versions of SE.Redis, and adds a lot of unnecessary complexity for a sample.

File: tutorial/output-cache-open-ai/infra/abbreviations.json (hunk 3 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/infra/abbreviations.json:1]: it's a generated file used by azd cli and bicep templates to create resource names without conflicting with other resources. It's optional but used for convenience. For example in the main.bicep you can see:
logAnalyticsName: '${abbrs.operationalInsightsWorkspaces}${resourceToken}'

File: tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs (hunk 2 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs:16]: fixed. will include in the next commit

File: tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs (hunk 1 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/OutputCacheDallESample/RedisConnection.cs:1]: deleted

File: tutorial/output-cache-open-ai/README.md (hunk 4 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/README.md:4]: fixed. Will include in the next commit

File: tutorial/output-cache-open-ai/README.md (hunk 6 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/README.md:37]: fixed. Will include in the next commit

File: tutorial/output-cache-open-ai/README.md (hunk 7 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/README.md:42]: Fixed. Yes registration is needed. Dall-E also has limited region availability. Will include in the next commit.

File: tutorial/output-cache-open-ai/README.md (hunk 5 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/README.md:9]: fixed. will include in the next commit.

File: tutorial/output-cache-open-ai/OutputCacheDallESample/Program.cs (hunk 9 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/OutputCacheDallESample/Program.cs:18]: fixed. will include in the next commit.

File: tutorial/output-cache-open-ai/infra/shared/apps-env.bicep (hunk 8 above)
Comment (CawaMS) [reply tutorial/output-cache-open-ai/infra/shared/apps-env.bicep:2]: @yugangw-msft  I don't think the infra folder should be shared - it's meant for a per-project deployment. i.e. anyone copying over a folder would be able to run the project with the bicep files. 
Let me know if I misunderstood anything.

**Comment by CawaMS** [comment]
@yugangw-msft  @philon-msft  I checked in all the fixes with the latest commit.

File: tutorial/output-cache-open-ai/infra/shared/apps-env.bicep (hunk 8 above)
Comment (yugangw-msft) [reply tutorial/output-cache-open-ai/infra/shared/apps-env.bicep:2]: I meant the bicep files to create a Redis Enterprise inside a VNET will be same across all the templates, correct? Efforts are worthwhile to consider so to reduce the duplicate to make sure future update will be simpler.

File: tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                    // Image Generations responses provide URLs you can use to retrieve requested images
+                    imageURL = imageGenerations.Value.Data[0].Url.AbsoluteUri;
+
+                    await cache.StoreAsync(_prompt, imageURL);
+
+                    await context.Response.WriteAsync("<!DOCTYPE html><html><body> " +
+                    $"<img src=\"{imageURL}\" alt=\"AI Generated Picture {_prompt}\" width=\"460\" height=\"345\">" +
+                    " </body> </html>");
+                }
+            }
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs:63]: remove empty lines below

File: tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
+﻿using Azure.AI.OpenAI;
+using Azure;
```
Comment (yugangw-msft) [inline tutorial/output-cache-open-ai/OutputCacheDallESample/GenerateImageSC.cs:2]: remove unused `using`

File: tutorial/output-cache-open-ai/README.md (hunk 5 above)
Comment (yugangw-msft) [reply tutorial/output-cache-open-ai/README.md:9]: Let us call out user need to update the url in browser to include the `nocache` endpoint

**Comment by CawaMS** [comment]
@yugangw-msft fixed in the latest commit

**Review by yugangw-msft — APPROVED** [review]
